exporter/sumologicexporter/exporter.go

```diff
@@ -15,45 +15,296 @@
-
 func newLogsExporter(
-	logger *zap.Logger,
+	cfg *Config,
 ) (component.LogsExporter, error) {
-	se := &sumologicexporter{logger: logger}
-	return se, nil
+	se, err := initExporter(cfg)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return exporterhelper.NewLogsExporter(
+		cfg,
+		se.pushLogsData,
+		// Disable exporterhelper Timeout, since we are using a custom mechanism
+		// within exporter itself
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
+		exporterhelper.WithRetry(cfg.RetrySettings),
+		exporterhelper.WithQueue(cfg.QueueSettings),
+	)
 }
 
-func newMetricsExporter(
-	logger *zap.Logger,
-) (component.MetricsExporter, error) {
-	se := &sumologicexporter{logger: logger}
+func initExporter(cfg *Config) (*sumologicexporter, error) {
+	se := &sumologicexporter{
+		config: cfg,
+		client: &http.Client{
+			Timeout: cfg.TimeoutSettings.Timeout,
+		},
+	}
+	err := se.refreshMetadataRegexes()
+
+	if err != nil {
+		return nil, err
+	}
+
 	return se, nil
 }
 
-func (se *sumologicexporter) Start(_ context.Context, _ component.Host) error {
+func (se *sumologicexporter) refreshMetadataRegexes() error {
+	cfg := se.config
+	metadataRegexes := make([]*regexp.Regexp, len(cfg.MetadataFields))
+	for i := 0; i < len(cfg.MetadataFields); i++ {
+		regex, err := regexp.Compile(cfg.MetadataFields[i])
+		if err != nil {
+			return err
+		}
+		metadataRegexes[i] = regex
+	}
+
+	se.metadataRegexes = metadataRegexes
 	return nil
 }
 
-func (se *sumologicexporter) Shutdown(context.Context) error {
+// filterMetadata returns map of attributes which are (or are not, it depends on filterOut argument) metadata
+func (se *sumologicexporter) filterMetadata(attributes pdata.AttributeMap, filterOut bool) map[string]string {
+	returnValue := make(map[string]string)
+	attributes.ForEach(func(k string, v pdata.AttributeValue) {
+		skip := !filterOut
+		for j := 0; j < len(se.metadataRegexes); j++ {
+			if se.metadataRegexes[j].MatchString(k) {
+				skip = false
+				if filterOut {
+					return
+				}
+			}
+		}
+
+		if skip {
+			return
+		}
+
+		returnValue[k] = v.StringVal()
+	})
+
+	return returnValue
+}
+
+// GetMetadata builds string which represents metadata in alphabetical order
+func (se *sumologicexporter) GetMetadata(attributes pdata.AttributeMap) string {
+	attrs := se.filterMetadata(attributes, false)
+	metadata := make([]string, 0, len(attrs))
+
+	for k := range attrs {
+		metadata = append(metadata, fmt.Sprintf("%s=%s", k, attrs[k]))
+	}
+	sort.Strings(metadata)
+
+	return strings.Join(metadata, ", ")
+}
+
+// This function tries to send data and eventually appends error in case of failure
+// It modifies buffer, droppedTimeSeries and errs
+func (se *sumologicexporter) sendAndPushErrors(buffer *[]pdata.LogRecord, fields string, droppedTimeSeries *int, errs *[]error) {
+	err := se.sendLogs(*buffer, fields)
+	if err != nil {
+		*droppedTimeSeries += len(*buffer)
+		*errs = append(*errs, err)
+	}
+	*buffer = (*buffer)[:0]
+}
```

Comment:
The point was to avoid copying as much code as possible. I can change it, but every usage of the `sendAndPushErrors` will require those 3 additional lines